From 986ec16d2be427c482e718ea0f2ec399b021aca5 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Mon, 20 May 2024 11:53:51 +0200
Subject: [PATCH 1130/1141] rpmsg: i2c: fix completion re-initialization

Some strange timeout has been observed, so take care, to initialize
completion upon probe, and call reinit_completion() at runtime.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I6546a47b5f2ab21bd76c623975a3b39889d53c19
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/380944
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/i2c/busses/i2c-rpmsg.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/i2c/busses/i2c-rpmsg.c b/drivers/i2c/busses/i2c-rpmsg.c
index 9f1bb4435a5b..33f59975c4d0 100644
--- a/drivers/i2c/busses/i2c-rpmsg.c
+++ b/drivers/i2c/busses/i2c-rpmsg.c
@@ -124,7 +124,7 @@ static int rpmsg_i2c_write(struct rpmsg_i2c_dev *ri2c_dev, struct i2c_msg *msg)
 
 	dev_dbg(&rpdev->dev, "%s: %x (len=%d)", __func__, r_msg->addr, r_msg->count);
 	ri2c_dev->is_write = true;
-	init_completion(&ri2c_dev->complete);
+	reinit_completion(&ri2c_dev->complete);
 
 	ret = rpmsg_send(rpdev->ept, r_msg, msg_size);
 	if (ret) {
@@ -161,7 +161,7 @@ static int rpmsg_i2c_read(struct rpmsg_i2c_dev *ri2c_dev, struct i2c_msg *msg)
 	r_msg->result = 0;
 
 	ri2c_dev->is_read = true;
-	init_completion(&ri2c_dev->complete);
+	reinit_completion(&ri2c_dev->complete);
 
 	dev_dbg(&rpdev->dev, "%s: %x (len=%d)", __func__, r_msg->addr, r_msg->count);
 
@@ -291,6 +291,7 @@ static int rpmsg_i2c_probe(struct rpmsg_device *rpdev)
 	adap->algo = &rpmsg_i2c_algo;
 	adap->dev.parent = ri2c_dev->dev;
 	adap->dev.of_node = ri2c_dev->dev->of_node;
+	init_completion(&ri2c_dev->complete);
 
 	/* match between proc id and service name */
 	dev_info(&rpdev->dev, "new channel: 0x%x -> 0x%x!\n",
-- 
2.39.2

